File: problems/C++/36. Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<unordered_set<char>> rows(9);
        vector<unordered_set<char>> cols(9);
        vector<unordered_set<char>> boxs(9);

        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                char val = board[r][c];
                if (val == '.') {
                    continue;
                }

                int b_idx = (r / 3) * 3 + (c / 3);
                
                if (rows[r].count(val) || cols[c].count(val) || boxs[b_idx].count(val)) {
                    return false;
                }
                
                rows[r].insert(val);
                cols[c].insert(val);
                boxs[b_idx].insert(val);
            }
        }
        return true;
    }
};

File: problems/Python/36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxs = [set() for _ in range(9)]

        for r in range(9):
            for c in range(9):
                val = board[r][c]
                if val == '.':
                    continue
                
                b_idx = (r // 3) * 3 + (c // 3)

                if val in rows[r] or val in cols[c] or val in boxs[b_idx]:
                    return False
                
                rows[r].add(val)
                cols[c].add(val)
                boxs[b_idx].add(val)
        
        return True